### PR TITLE
[SPARK-19637][SQL] Add to_json in FunctionRegistry

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added entries  in `FunctionRegistry` and supported `to_json` in SQL.

## How was this patch tested?
Added tests in `JsonFunctionsSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 1]
```diff
@@ -18,7 +18,14 @@
 package org.apache.spark.sql.catalyst.json
 
 import com.fasterxml.jackson.core.{JsonParser, JsonToken}
+import org.json4s._
+import org.json4s.JsonAST.JValue
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: This is just my personal opinion but should we maybe consider minimising this import? For example, `import org.json4s.jackson.JsonMethods.parse`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 2]
```diff
@@ -55,4 +62,32 @@ object JacksonUtils {
...
+    DataType.fromJson(validateStringLiteral(exp)).asInstanceOf[StructType]
+
+  /**
+   * Convert a literal including a json option string (e.g., '{"mode": "PERMISSIVE", ...}')
+   * to Map-type data.
+   */
+  def validateOptionsLiteral(exp: Expression): Map[String, String] = {
+    val json = validateStringLiteral(exp)
+    parse(json) match {
+      case JObject(options) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: We could try to utilize `val parse(json).extract[Map[String, String]]`. Given my observation, it produces empty `Map` if there are no such values or empty json and it throws an exception if it is an invalid json.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 3]
```diff
@@ -174,4 +174,44 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
...
+    val df2 = Seq(Tuple1(Tuple1(java.sql.Timestamp.valueOf("2015-08-26 18:00:00.0")))).toDF("a")
+    checkAnswer(
+      df2.selectExpr("""to_json(a, '{"timestampFormat": "dd/MM/yyyy HH:mm"}')"""),
+      Row("""{"_1":"26/08/2015 18:00"}""") :: Nil)
+
+    val errMsg1 = intercept[AnalysisException] {
+      df2.selectExpr("""to_json(a, '{"k": [{"k": "v"}]}')""").collect
+    }
+    assert(errMsg1.getMessage.startsWith(
+      "The format must be '{\"key\": \"value\", ...}', but {\"k\": [{\"k\": \"v\"}]}"))
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Maybe, `""" ... """` if more commits should be pushed.

**Review by HyukjinKwon — COMMENTED** [review]
@maropu, I just left few opinions that might help.

**Comment by maropu** [comment]
@HyukjinKwon Thanks! I'll check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 4]
```diff
@@ -55,4 +60,26 @@ object JacksonUtils {
...
+   * Convert a literal including a json option string (e.g., '{"mode": "PERMISSIVE", ...}')
+   * to Map-type data.
+   */
+  def validateOptionsLiteral(exp: Expression): Map[String, String] = {
+    implicit val formats = org.json4s.DefaultFormats
+    val json = validateStringLiteral(exp)
+    Try(parse(json).extract[Map[String, String]]) match {
+      case Success(m) => m
+      case Failure(_) =>
+        throw new AnalysisException(s"""The format must be '{"key": "value", ...}', but ${json}"""")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Could we do this as below

```scala
    Try(parse(json).extract[Map[String, String]]).getOrElse {
      throw new AnalysisException(...)
    }
```

or maybe just a try-catch block?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 5]
```diff
@@ -55,4 +60,26 @@ object JacksonUtils {
 
     schema.foreach(field => verifyType(field.name, field.dataType))
   }
+
+  private def validateStringLiteral(exp: Expression): String = exp match {
+    case Literal(s, StringType) => s.toString
+    case e => throw new AnalysisException("Must be a string literal, but: " + e)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: trivial..

```
s"Must be a string literal, but: $e"
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 6]
```diff
@@ -55,4 +60,26 @@ object JacksonUtils {
...
+  }
+
+  def validateSchemaLiteral(exp: Expression): StructType =
+    DataType.fromJson(validateStringLiteral(exp)).asInstanceOf[StructType]
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: I guess we should check if it is `StructType` and throw a proper exception because it seems `JsonToStruct` does not check if `exp` is `StructType` and it probably throws a cast exception (if I haven't missed something here).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 7]
```diff
@@ -482,6 +482,15 @@ case class JsonTuple(children: Seq[Expression])
 /**
  * Converts an json input string to a [[StructType]] with the specified schema.
  */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(jsonStr, schema[, options]) - Return a `StructType` value with the given `jsonStr` and `schema`.",
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Probably just `struct` instead of `` `StructType` `` (as I found a example in https://github.com/apache/spark/blob/e7f982b20d8a1c0db711e0dcfe26b2f39f98dd64/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala#L340 as a reference).

**Review by HyukjinKwon — COMMENTED** [review]
Sorry, I happened to review twice. These are all from me. I hope my comments sound reasonable. Other than them, (FWIW) it looks good to me.

**Comment by maropu** [comment]
I'll update in a day, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala (hunk 6 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: I just wrote this way along with here https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L3010. Both is okay to me though, if we modify the code in a way you suggested, we need to modify `from_json` code, too?

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Ah, thanks. Yes, if it throws a class cast exception, I think we should produce a better exception and message rather than just one saying `A cannot be cast to B`. Maybe, add a util for both places?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: okay, I'll do that ;)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 8]
```diff
@@ -482,6 +482,15 @@ case class JsonTuple(children: Seq[Expression])
 /**
  * Converts an json input string to a [[StructType]] with the specified schema.
  */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(jsonStr, schema[, options]) - Return a struct value with the given `jsonStr` and `schema`.",
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: `Return` -> `Returns`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 9]
```diff
@@ -55,4 +60,24 @@ object JacksonUtils {
...
+  private def validateStringLiteral(exp: Expression): String = exp match {
+    case Literal(s, StringType) => s.toString
+    case e => throw new AnalysisException(s"Must be a string literal, but: $e")
+  }
+
+  def validateSchemaLiteral(exp: Expression): StructType =
+    DataType.fromJson(validateStringLiteral(exp)).asInstanceOf[StructType]
+
+  /**
+   * Convert a literal including a json option string (e.g., '{"mode": "PERMISSIVE", ...}')
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: What is the reason we use the Json option string?

**Comment by gatorsmile** [comment]
Could you add SQL test cases to SQLQueryTestSuite?

**Comment by maropu** [comment]
@gatorsmile okay, I'll do soon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 10]
```diff
@@ -482,6 +482,15 @@ case class JsonTuple(children: Seq[Expression])
 /**
  * Converts an json input string to a [[StructType]] with the specified schema.
  */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(jsonStr, schema[, options]) - Returns a struct value with the given `jsonStr` and `schema`.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('{"a":1}', '{"type":"struct", "fields":[{"name":"a", "type":"integer", "nullable":true}]}');
+       {"a":1}
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: More examples are needed to show users how to use option.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala (hunk 9 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Aha, you mean we use a map literal, directly? Sorry, but I missed that idea. This json option is totally meaningless? If yes, I'll fix to use a map literal here.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 11]
```diff
@@ -174,4 +174,44 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
...
+    checkAnswer(
+      df3.selectExpr(s"from_json(value, '${schema1.json}')"),
+      Row(Row(1)) :: Nil)
+
+    val df4 = Seq("""{"time": "26/08/2015 18:00"}""").toDS()
+    val schema2 = new StructType().add("time", TimestampType)
+    checkAnswer(
+      df4.selectExpr(
+        s"""from_json(value, '${schema2.json}', """ +
+           """'{"timestampFormat": "dd/MM/yyyy HH:mm"}')"""),
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Regarding the format of options, another way is to use the MapType.

For example, ```Scala
from_json(value, '${schema2.json}', map("timestampFormat", "dd/MM/yyyy HH:mm"))
```

I am not sure whether using JSON to represent options is a good way.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 12]
```diff
@@ -174,4 +174,44 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
...
+      df1.selectExpr("to_json(a)"),
+      Row("""{"_1":1}""") :: Nil)
+
+    val df2 = Seq(Tuple1(Tuple1(java.sql.Timestamp.valueOf("2015-08-26 18:00:00.0")))).toDF("a")
+    checkAnswer(
+      df2.selectExpr("""to_json(a, '{"timestampFormat": "dd/MM/yyyy HH:mm"}')"""),
+      Row("""{"_1":"26/08/2015 18:00"}""") :: Nil)
+
+    val errMsg1 = intercept[AnalysisException] {
+      df2.selectExpr("""to_json(a, '{"k": [{"k": "v"}]}')""").collect
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: `collect ` is not needed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 13]
```diff
@@ -482,6 +482,15 @@ case class JsonTuple(children: Seq[Expression])
 /**
  * Converts an json input string to a [[StructType]] with the specified schema.
  */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(jsonStr, schema[, options]) - Returns a struct value with the given `jsonStr` and `schema`.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('{"a":1}', '{"type":"struct", "fields":[{"name":"a", "type":"integer", "nullable":true}]}');
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Can we let users call `named_struct` function to specify the schema?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I'll check

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 11 above)
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: okay, I'll fix in that way.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 13 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: @gatorsmile, do you mind if I ask to elaborate what you think wtih `named_struct`? I am just curious.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I checked related code though, if we use `named_struc` here, we need to add substantial code to convert `named_struct` to `StructType`...

**Comment by maropu** [comment]
ping

**Comment by maropu** [comment]
@gatorsmile ping

**Comment by maropu** [comment]
@gatorsmile ping

**Comment by gatorsmile** [comment]
> From JSON is harder because the second argument is a StructType. We could consider accepting a string in the DDL format for declaring a tables schema (i.e. a: Int, b: struct<a:Int, c: String>....

To parse the schema represented in DDL format, instead of the json format, we need to call the parser to do it. If you are not familar with the parser, maybe you only implement `to_json` in this PR?

**Comment by maropu** [comment]
So, this pr re-used `DataType.fromJson` here: https://github.com/apache/spark/pull/16981/files#diff-113a2b8242f0ee6ec3914f539f119619R65. But, I know this is some arguable. I also think it'd be better to drop off `from_json` from this pr and make a new JIRA to discuss `from_json`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 14]
```diff
@@ -55,4 +60,22 @@ object JacksonUtils {
 
     schema.foreach(field => verifyType(field.name, field.dataType))
   }
+
+  def strToStructType(schemaAsJson: String): StructType = Try {
+    DataType.fromJson(schemaAsJson).asInstanceOf[StructType]
+  }.getOrElse {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: > prefer explicitly throwing exceptions for abnormal execution and Java style try/catch for exception handling.

Above is the style recommendation. Please use try/catch

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 15]
```diff
@@ -55,4 +60,22 @@ object JacksonUtils {
...
+    throw new AnalysisException(
+      s"""Illegal json string for representing a schema: $schemaAsJson"""")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: \"\"\"\" Four?

`s"Illegal json string for representing a schema: $schemaAsJson"`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 16]
```diff
@@ -55,4 +60,22 @@ object JacksonUtils {
 
     schema.foreach(field => verifyType(field.name, field.dataType))
   }
+
+  def strToStructType(schemaAsJson: String): StructType = Try {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Maybe we do not need it.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 17]
```diff
@@ -3007,7 +3008,7 @@ object functions {
    * @since 2.1.0
    */
   def from_json(e: Column, schema: String, options: java.util.Map[String, String]): Column =
-    from_json(e, DataType.fromJson(schema).asInstanceOf[StructType], options)
+    from_json(e, JacksonUtils.strToStructType(schema), options)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Revert it back?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 18]
```diff
@@ -174,4 +174,22 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
       .select(to_json($"struct").as("json"))
     checkAnswer(dfTwo, readBackTwo)
   }
+
+  test("SPARK-19637 Support to_json in SQL") {
+    // to_json
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Nit: remove this comment.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 17 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: okay, I'll do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala (hunk 14 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: okay, I'll fix

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala (hunk 16 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: yes, I'll remove

**Comment by maropu** [comment]
@gatorsmile okay, I'll fix the issues you mentioned.

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 19]
```diff
@@ -55,4 +58,15 @@ object JacksonUtils {
...
     schema.foreach(field => verifyType(field.name, field.dataType))
   }
+
+  /** Convert a map literal to Map-type data. */
+  def validateMapData(exp: Expression): Map[String, String] = exp match {
+    case m: CreateMap => m.dataType.acceptsType(MapType(StringType, StringType, false))
+      val arrayMap = m.eval().asInstanceOf[ArrayBasedMapData]
+      ArrayBasedMapData.toScalaMap(arrayMap).map { case (key, value) =>
+        key.toString -> value.toString
+      }.toMap
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: `.toMap` is needed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 20]
```diff
@@ -55,4 +58,15 @@ object JacksonUtils {
 
     schema.foreach(field => verifyType(field.name, field.dataType))
   }
+
+  /** Convert a map literal to Map-type data. */
+  def validateMapData(exp: Expression): Map[String, String] = exp match {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Why we need to keep here? This is not related to `JacksonUtils.scala`. The function name also needs a change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 21]
```diff
@@ -55,4 +58,15 @@ object JacksonUtils {
...
+    case m: CreateMap => m.dataType.acceptsType(MapType(StringType, StringType, false))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]:
```
case m: CreateMap if m.dataType.acceptsType(MapType(StringType, StringType, false)) =>
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: Nit: `valueContainsNull = false`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 22]
```diff
@@ -566,3 +586,17 @@ case class StructToJson(
...
+
+  def convertToMapData(exp: Expression): Map[String, String] = exp match {
+    case m: CreateMap
+        if m.dataType.acceptsType(MapType(StringType, StringType, valueContainsNull = false)) =>
+      val arrayMap = m.eval().asInstanceOf[ArrayBasedMapData]
+      ArrayBasedMapData.toScalaMap(arrayMap).map { case (key, value) =>
+        key.toString -> value.toString
+      }
+    case _ =>
+      throw new AnalysisException("Must use a map() function for options")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: This message is misleading in the following case:
```
df2.selectExpr("to_json(a, map('a', 1))")
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Please also include the test case for this. Thanks!

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: okay!

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: How about this?
https://github.com/apache/spark/pull/16981/files#diff-6626026091295ad8c0dfb66ecbcd04b1R601

**Comment by gatorsmile** [comment]
LGTM 

cc @marmbrus @brkyvz Is this impl what you expect?

**Comment by marmbrus** [comment]
yeah, LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by gatorsmile** [comment]
cc @maropu https://github.com/apache/spark/pull/17171 is merged. Are you interested in working on `from_json`?

JIRA: https://issues.apache.org/jira/browse/SPARK-19967
